clippy_lints/src/pathbuf_init_then_push.rs

```diff
@@ -0,0 +1,151 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::path_to_local_id;
+use clippy_utils::source::{snippet, snippet_opt};
+use clippy_utils::ty::is_type_diagnostic_item;
+use rustc_errors::Applicability;
+use rustc_hir::def::Res;
+use rustc_hir::{BindingAnnotation, Block, Expr, ExprKind, HirId, Local, PatKind, QPath, Stmt, StmtKind};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::{sym, Span, Symbol};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for calls to `push` immediately after creating a new `PathBuf`.
+    ///
+    /// ### Why is this bad?
+    /// The `.join()` is easier to read than multiple `push` calls.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let mut path_buf = PathBuf::new();
+    /// path_buf.push("foo");
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let path_buf = PathBuf::new().join("foo");
```

Comment:
Why wouldn't you use `PathBuf::from` here? We should lint both of these and suggest it instead, imo.

I suppose this would insert `/` or `\\` automatically but that's all I can really think of.

Reply:
I think this is very rare
Moreover, I'm not familar with the details of pathbuf

Reply:
I doubt it's not trial after reading https://github.com/rust-lang/rust-clippy/pull/11453

Reply:
wdym? I'm not sure why `PathBuf::new().join("foo").join("bar")` can't be `PathBuf::from("foo/bar")` (`/` is crossplatform iirc)

Reply:
I don't like overeager suggestion

Reply:
Now it would make sense to change this to `PathBuf::from` since that's an actual suggestion of the lint.

You can add two suggestions in the code description. Just have a comment like:

```
// Use PathBuf::from:
...

// You can also use `.join()`
...
```

Reply:
@lengyijun have you seen this comment?